projects/lexical_test/lib/lexical/test/cursor_support.ex

```diff
@@ -1,49 +1,98 @@
 defmodule Lexical.Test.CursorSupport do
-  def cursor_position(text) do
-    text
-    |> String.graphemes()
-    |> Enum.chunk_every(2, 1, [""])
-    |> Enum.reduce_while({starting_line(), starting_column()}, fn
-      ["|", ">"], {line, column} ->
-        {:cont, {line, column + 1}}
+  @moduledoc """
+  Utilities for extracting cursor position in code fragments and documents.
+  """
 
-      ["|", _], position ->
-        {:halt, position}
+  alias Lexical.Document
+  alias Lexical.Document.Position
+  alias Lexical.Test.PositionSupport
 
-      ["\n", _], {line, _column} ->
-        {:cont, {line + 1, starting_column()}}
+  @default_cursor "|"
+  @starting_line 1
+  @starting_column 1
 
-      _, {line, column} ->
-        {:cont, {line, column + 1}}
-    end)
-  end
+  @type cursor_position :: {pos_integer(), pos_integer()}
 
-  def context_before_cursor(text) do
-    text
-    |> String.graphemes()
-    |> Enum.chunk_every(2, 1, [""])
-    |> Enum.reduce_while([], fn
-      ["|", ">"], iodata ->
-        {:cont, [iodata, "|"]}
+  @doc """
+  Finds a cursor in `text` and returns a tuple of the cursor position and
+  the text with the cursor stripped out.
 
-      ["|", _lookahead], iodata ->
-        {:halt, iodata}
+  ## Options
 
-      [c, _], iodata ->
-        {:cont, [iodata, c]}
-    end)
-    |> IO.iodata_to_binary()
+    * `:cursor` - the cursor string to be found. Defaults to
+      `#{inspect(@default_cursor)}`.
+
+    * `:as` - one of `:text` (default) or `:document`. If `:document`,
+      wraps the text in a `Lexical.Document` using the URI `"file:///file.ex"`.
+
+    * `:document` - the document path or URI. Setting this option implies
+      `as: :document`.
+
+  ## Examples
+
+      iex> code = \"""
+      ...> defmodule MyModule do
+      ...>   alias Foo|
+      ...> end
+      ...> \"""
+
+      iex> pop_cursor(code)
+      {
+        %Position{line: 2, column: 12},
+        \"""
+        defmodule MyModule do
+          alias Foo
+        end
+        \"""
+      }
+
+      iex> pop_cursor(code, as: :document)
+      {
+        %Position{line: 2, column: 12},
+        %Document{uri: "file:///file.ex", ...}
+      }
+
+      iex> pop_cursor(code, document: "my_doc.ex")
+      {
+        %Position{line: 2, column: 12},
+        %Document{uri: "file:///my_doc.ex", ...}
+      }
+
+  """
+  @spec pop_cursor(text :: String.t(), [opt]) :: {Position.t(), String.t() | Document.t()}
+        when opt: {:cursor, String.t()} | {:as, :text | :document} | {:document, String.t()}
+  def pop_cursor(text, opts \\ []) do
```

Comment:
I'm like this change; `pop_cursor` returns `{cursor, text_without_cursor}` really makes sense.